src/hotspot/share/runtime/handshake.cpp

```diff
@@ -562,3 +569,66 @@ HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* ma
                        pr_ret == HandshakeState::_succeeded ? "including" : "excluding", p2i(match_op));
   return pr_ret;
 }
+
+void HandshakeState::lock() {
+  _lock.lock_without_safepoint_check();
+}
+
+void HandshakeState::unlock() {
+  _lock.unlock();
+}
+
+// The careful dance between thread suspension and exit is handled here.
+// The caller has notified the agents that we are exiting, before we go
+// on, we must check for a pending external suspend request and honor it
+// in order to not surprise the thread that made the suspend request.
+void HandshakeState::thread_exit() {
+  assert(JavaThread::current() == _handshakee, "Must be");
+  while (true) {
+    {
+      MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+      if (!_handshakee->is_suspend_requested()) {
+        _handshakee->set_exiting();
+        // No more external suspends are allowed at this point.
+        return;
+      }
+    }
+    ThreadBlockInVM tbivm(_handshakee);
+    // We're done with this suspend request, but we have to loop around
+    // and check again. Eventually we will get the lock without a pending
+    // external suspend request and will be able to mark ourselves as
+    // exiting.
+  }
+}
+
+void HandshakeState::suspend_in_handshake() {
+  assert(Thread::current() == _handshakee, "should call from _handshakee");
```

Comment:
This one check is critical for correctness of the new Handshake based
suspend mechanism. I'd be tempted to make this one assert() into a
guarantee() while we shake out this new code.

Reply:
Sure!